tools/clang/include/clang/SPIRV/SpirvInstruction.h

```diff
@@ -241,6 +244,7 @@ class SpirvInstruction {
   bool containsAlias;
 
   spv::StorageClass storageClass;
+  spv::ImageFormat imageFormat;
```

Comment:
Based on the [SPIR-V spec](https://www.khronos.org/registry/spir-v/specs/unified1/SPIRV.html#_a_id_image_format_a_image_format), the ImageFormat is only used by the Image **Type**. (no SPIR-V instruction directly uses the image format).

Therefore putting this information inside the `SpirvInstruction` class may not be the right thing to do. Ideally it should be placed inside the `ImageType` class. In fact, this class [already has a member](https://github.com/microsoft/DirectXShaderCompiler/blob/4ced9bdf05ba26c6d9c01258ad8914f892e3620a/tools/clang/include/clang/SPIRV/SpirvType.h#L214) for the image format. It just needs to be populated correctly.

The fact that you had to add `// Propagate image format to the result type of load` code to manually propagate the information after OpLoad is a side effect of this. (ideally you shouldn't need to have a special case code for OpLoad. Are there other special cases that are missed?).

Having said all of this, the place where we set the image format in the image type is in the LowerTypeVisitor, and when we reach LowerTypeVisitor, we have already lost the `vk::image_format` attribute because we don't have access to Decls inside LowerTypeVisitor. @jaebaek, please think about this more carefully and provide a solution.

---
P.S. If we think putting this information inside the SpirvInstruction is the way to go, the propagation of the image format in the types should be done in a separate visitor pass before LowerTypeVisitor.

Reply:
Regarding ImageType, the format is already put there: